KAFKA-15415: On producer-batch retry, skip-backoff on a new leader

msn-tldr opened the pull request:
When producer-batch is being retried, new-leader is known for the partition Vs the leader used in last attempt, then it is worthwhile to retry immediately to this new leader. A partition-leader is considered to be newer, if the epoch has advanced.

Code change summary -

1. `ProducerBatch` now tracks latest leader-epoch on which attempts will be made, and change in leader(epoch) is tracked through method `hasLeaderChanged`.
2. `RecordAccumulator` decides if batch should immediately-retry(skip-backoff), or backoff, on detecting a leader change.
3. All classes have UTs & e2e test in `SenderTest.java`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

msn-tldr committed:
KAFKA-15415 For ProducerBatch being retried, skip-backoff to a new leader

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/PartitionInfo.java:
```diff
@@ -20,26 +20,35 @@
     private final String topic;
     private final int partition;
     private final Node leader;
+    private final int leaderEpoch;
```
I think it's more aligned with the existing Kafka interface to use `Optional<Integer> leaderEpoch`. For example, `org.apache.kafka.clients.consumer.OffsetAndMetadata`.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/PartitionInfo.java:
```diff
@@ -20,26 +20,35 @@
 
     public PartitionInfo(String topic, int partition, Node leader, Node[] replicas, Node[] inSyncReplicas) {
-        this(topic, partition, leader, replicas, inSyncReplicas, new Node[0]);
+        this(topic, partition, leader, UNKNOWN_LEADER_EPOCH, replicas, inSyncReplicas, new Node[0]);
```
And then `Optional.empty()` of course.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/PartitionInfo.java:
```diff
@@ -60,12 +69,19 @@ public int partition() {
     }
 
+    /**
+     * The epoch of the partition's leader.
```
I think it's the "The leader epoch of this partition if known".

kirktrue replied:
`Optional` here, right?

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/Cluster.java:
```diff
@@ -277,6 +277,19 @@ public Node leaderFor(TopicPartition topicPartition) {
+     * @param topicPartition
+     * @return The epoch for partition's leader, or UNKNOWN_LEADER_EPOCH if epoch unkown.
+     */
+    public int leaderEpochFor(TopicPartition topicPartition) {
```
Again, I think `Optional<Integer>` is more usual.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/Cluster.java:
```diff
@@ -277,6 +277,19 @@ public Node leaderFor(TopicPartition topicPartition) {
+    /**
+     * Get the current leader's epoch for the given topic-partition.
+     * @param topicPartition
+     * @return The epoch for partition's leader, or UNKNOWN_LEADER_EPOCH if epoch unkown.
```
Typo. "unkown" -> "unknown".

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
+    boolean hasLeaderChanged(int latestLeaderEpoch) {
+        boolean leaderChanged = false;
+        // Checking for leader change makes sense only from 1st retry onwards(attempt >=1).
+        log.trace("For {}, attempting to change leader, currentLeaderEpoch:{}, leaderChangedAttempts:{}, latestLeaderEpoch: {}, current Attempt: {}",
```
There are 4 separate calls to `attempts()` (or the equivalent of `attempts.get()`) in this method. Capture the value with a single call and use instead.

kirktrue replied:
The value shouldn't change between those calls, right?

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
         float compressionRatioEstimation = CompressionRatioEstimator.estimation(topicPartition.topic(),
                                                                                 recordsBuilder.compressionType());
+        this.currentLeaderEpoch = PartitionInfo.UNKNOWN_LEADER_EPOCH;
+        this.leaderChangedAttempts = -1;
```
I don't think there's any need for a special value of -1. Initializing to 0 would work fine.

kirktrue replied:
Seems as though 0 would do, yes.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -79,6 +80,11 @@ private enum FinalState { ABORTED, FAILED, SUCCEEDED }
+    // Tracks the current-leader's epoch to which this batch would be sent, in the current to produce the batch.
+    private int currentLeaderEpoch;
+    // Tracks the attempt in which leader was changed to currentLeaderEpoch for the 1st time.
+    private int leaderChangedAttempts;
```
Sounds like the number of attempts to change the leader. Perhaps `attemptsWhenLeaderLastChanged` would be more descriptive.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
+            }
+        }
+        if (leaderChanged) {
+            log.debug("For {}, leaderChanged, currentLeaderEpoch:{}, leaderChangedAttempts:{}",
```
I don't see why you're tracing only the previous leader epoch here. I think you're interested in the new one, or both new and old. Because the assignment follows the log line, the "current" is still previous.

kirktrue replied:
Agreed.

kirktrue replied:
```suggestion
            log.debug("For {}, a leader change was detected, currentLeaderEpoch: {}, leaderChangedAttempts: {}",
```

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -517,4 +554,14 @@ public boolean isTransactional() {
+    }
+
+    // VisibleForTesting
+    int leaderChangedAttempts() {
```
`attemptsWhenLeaderLastChanged()` perhaps.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -156,6 +157,8 @@ public class SenderTest {
     private SenderMetricsRegistry senderMetricsRegistry = null;
     private final LogContext logContext = new LogContext();
 
+    private final Logger log = logContext.logger(SenderTest.class);
```
I'm surprised that you added a logger when this massive test hasn't used one so far. Seems like logging is not required for working out whether the test passed or how it failed. I suggest removing it.

msn-tldr replied:
Inclined to keep it since helps improve readability of test-logs when test fails. By default logging is off, see similar comment below for details.

wcarlson5 replied:
Generally adding logging in tests isn't great. I would second removing it. I don't think the two lines added are worth the change. 

If you want that information about why a batch isn't being processed or if a batch is being processed you should add them as logs to the actual code. They can be trace if necessary

msn-tldr replied:
Ok fair enough will remove the logging from tests.

I have already added traces in actual code, so that should be fine for now.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -96,6 +98,8 @@ public class RecordAccumulatorTest {
     private final long maxBlockTimeMs = 1000;
     private final LogContext logContext = new LogContext();
 
+    private final Logger log = logContext.logger(RecordAccumulatorTest.class);
```
Probably unnecessary.

kirktrue replied:
I am not personally against logging in tests, but they should be turned down so that they're not emitted by default.

msn-tldr replied:
I am inclined to keep it, as it helps improve readability of test logs if/when test fails. 
By default, the logging is turned off in `clients/src/resources/log4.properties`

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -3146,6 +3149,97 @@ public void testInvalidTxnStateIsAnAbortableError() throws Exception {
+            sender.runOnce(); // send produce request, immediately.
+            assertEquals(1, sender.inFlightBatches(tp0).size());
+            assertTrue(client.hasInFlightRequests());
+            client.respond(produceResponse(tp0, -1, Errors.NOT_LEADER_OR_FOLLOWER, 0));
```
Should there not be testing that uses the new `CurrentLeader` field in v9 of the ProduceResponse given that this is part of KIP-951 which this PR is implementing?

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
+    boolean hasLeaderChanged(int latestLeaderEpoch) {
+        boolean leaderChanged = false;
+        // Checking for leader change makes sense only from 1st retry onwards(attempt >=1).
+        log.trace("For {}, attempting to change leader, currentLeaderEpoch:{}, leaderChangedAttempts:{}, latestLeaderEpoch: {}, current Attempt: {}",
```
Can I be extremely nit-picky about the capitalization of `Attempt` in the log line? Any reason it can't just be `attempt`?

kirktrue replied:
Also super nit-picky: spaces after colons in message.

Sorry, I can't help myself 😛

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
+     * @param latestLeaderEpoch The latest leader epoch.
+     * @return true if the leader has changed, otherwise false.
+     */
+    boolean hasLeaderChanged(int latestLeaderEpoch) {
```
```suggestion
    boolean maybeUpdateLeaderEpoch(int latestLeaderEpoch) {
```

This will imply that we're not just determining that the leader has changed, we're updating internal state.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
+                leaderChangedAttempts = attempts();
+                leaderChanged = true;
+            } else {
+                // Otherwise, it's only a leader change until the first attempt is made with this leader
```
I'm sorry, I don't quite get this logic.

If there are at least one attempt and the above epoch comparison is false, why do we consider the leader as having changed?

msn-tldr replied:
> least one attempt

implies a retry.

> the above epoch comparison is false

So leader-epoch is still the same. 

Consider leader was changed at attempt=5, to epoch=100. maybeUpdateLeaderEpoch() should detect leader change even when called again at attempt=5, with the same epoch=100. As this is the same attempt in which the leader change was detected.

kirktrue replied:
I guess I was under the false impression that `maybeUpdateLeaderEpoch()` was only called _once_ per attempt. I _think_ I can see how the flow of the code would end up calling that multiple times.

msn-tldr replied:
> I think I can see how the flow of the code would end up calling that multiple times.

I gets called multiple times. Its sendProducerData -> ready -> partitionReady, and then again in sendProducerData -> drain -> drainBatchesForOneNode.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -859,8 +863,9 @@ private List<ProducerBatch> drainBatchesForOneNode(Cluster cluster, Node node, i
 
             Deque<ProducerBatch> deque = getDeque(tp);
-            if (deque == null)
+            if (deque == null) {
```
Unless there's a valid reason to, we should omit formatting changes.

kirktrue commented on clients/src/main/java/org/apache/kafka/common/PartitionInfo.java:
```diff
@@ -20,26 +20,35 @@
     public PartitionInfo(String topic,
                          int partition,
                          Node leader,
+                         int leaderEpoch,
```
This is where we might get into trouble since we're breaking a constructor by adding a parameter.

kirktrue reviewed (changes requested):
Took a pass, except for tests.

@jsancio - can you weigh in about changes to `org.apache.kafka.common` needing a KIP?

msn-tldr committed:
minor refactor of PartitionInfo

msn-tldr committed:
Refactor to use metadata, Vs cluster, in RecordAccumulator

Using metadata gives leader-epoch info required to to check if leader has changed, for skipping backoff

dajac commented on clients/src/main/java/org/apache/kafka/common/PartitionInfo.java:
```diff
@@ -20,6 +20,7 @@
  * This is used to describe per-partition state in the MetadataResponse.
  */
 public class PartitionInfo {
+    public static final int UNKNOWN_LEADER_EPOCH = -1;
```
I think that `PartitionInfo` is a public class so I would not add this here.

msn-tldr replied:
@dajac this is removed now, thanks!

msn-tldr committed:
Address PR review comments

msn-tldr commented:
@kirktrue 
> @jsancio - can you weigh in about changes to org.apache.kafka.common needing a KIP?

I have removed the changes to public classes, so this shouldn't require a KIP, take a look.

wcarlson5 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -94,9 +100,45 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
+     * @param latestLeaderEpoch The latest leader epoch.
+     * @return true if the leader has changed, otherwise false.
+     */
+    boolean maybeUpdateLeaderEpoch(Optional<Integer> latestLeaderEpoch) {
```
It seems to be that this doesn't actually update the Leader Epoch, but rather it checks to see if the leader has changed since the last retry. Can we name it something more clear?

kirktrue replied:
It does assign the `currentLeaderEpoch` at the end, though, right?

msn-tldr replied:
So the method does 2 things
1. updates leader if a newer leader is supplied
2. also return has changed for the ongoing retry of the batch. 

Keeping them both in 1 method, is a bit confusing, as evident in the naming 😄 

So i have split it into 2 methods, to me code is more readable now.

wcarlson5 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -712,13 +714,14 @@ private long partitionReady(Cluster cluster, long nowMs, String topic,
 
                 waitedTimeMs = batch.waitedTimeMs(nowMs);
-                backingOff = shouldBackoff(batch, waitedTimeMs);
+                boolean hasLeaderChanged = batch.maybeUpdateLeaderEpoch(leaderAndEpoch.epoch);
```
See that method name and variable name don't really make sense together

wcarlson5 commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -1398,16 +1416,253 @@ public void testBuiltInPartitionerFractionalBatches() throws Exception {
+     * 2. change in leader hosting the partition.
+     */
+    @Test
+    public void testReadyAndDrainWhenABatchIsBeingRetried() throws InterruptedException {
```
I think this is just my unfamiliarity talking, but how are you checking if it is backing off? Is it just relying on these lines? 

```       
assertEquals(Optional.of(part1LeaderEpoch), batch.currentLeaderEpoch());
assertEquals(0, batch.attemptsWhenLeaderLastChanged());
```

msn-tldr replied:
So if  
1. batch should backoff - then both `ready` & `drain` would not have node1 & batch respectively in their output.
```
assertFalse(result.readyNodes.contains(node1), "Node1 is not ready");
assertTrue(batches.containsKey(node1.id()) && batches.get(node1.id()).isEmpty(),
                "No batches ready to be drained on Node1");
```
2. batch should not be backing off - then both `ready` & `drain` would have node1 & batch respectively in their output.
```
assertTrue(result.readyNodes.contains(node1), "Node1 is ready");
assertTrue(batches.containsKey(node1.id()) && batches.get(node1.id()).size() == 1, "Node1 has 1 batch ready & drained");
```

Other assertions are sanity checking auxillary conditions

wcarlson5 commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -1398,16 +1416,253 @@ public void testBuiltInPartitionerFractionalBatches() throws Exception {
+    }
+
+    @Test
+    public void testDrainWithANodeThatDoesntHostAnyPartitions() {
```
Is this just added to increase coverage? Thats fine, but I don't see how its related to the rest of the PR

msn-tldr replied:
`RecordAccumulator` `drain` now uses Metadata Vs Cluster. Metadata can change where as Cluster is immutable. Hence added tests w.r.t metadata changing(which weren't possible earlier)
1. node(broker) hosts partitions but partitions map to it change underneath
2. node(broker) no longer hosts any partitions.

wcarlson5 replied:
Awesome, that answer my question perfectly

wcarlson5 commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -3146,6 +3149,97 @@ public void testInvalidTxnStateIsAnAbortableError() throws Exception {
+            long totalSize = 1024 * 1024;
+            BufferPool pool = new BufferPool(totalSize, batchSize, metrics, time,
+                metricGrpName);
+            long retryBackoffMaxMs = 100L;
```
How difficult would it be to mock the backoff using a ticker? Too many times having to rely on timing in tests makes them become flaky. This seems like it would low probability here but we should keep it in mind

msn-tldr replied:
Fair point. The `SenderTest` already mocks `time` using `MockTime`, and that is manually advanced. Is that what u had in mind?

wcarlson5 replied:
That's good to know. My concern came from `time.sleep(2 * retryBackoffMaxMs);` down on line 3223. I'm not sure that is avoidable though

msn-tldr replied:
`time.sleep(2 * retryBackoffMaxMs);` that is how i ensure that tests waits more than backoff period before retrying.

wcarlson5 reviewed (commented):
Overall its looking good. I have a few questions, mostly for my own edification

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -94,9 +100,45 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
+            }
+        }
+        if (leaderChanged) {
+            log.debug("For {}, leader has changed, oldEpoch: {}, newEpoch: {}",
```
nit: per @wcarlson5's [comment above](https://github.com/apache/kafka/pull/14384/files#r1344314253), the log message might be read that the internal state of the batch was _changed_, rather than simply _detected_. I don't have suggestions on how to reword though. It's fine if you want to leave it as is.

msn-tldr replied:
fixed

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -682,8 +683,9 @@ private long partitionReady(Cluster cluster, long nowMs, String topic,
-            Node leader = cluster.leaderFor(part);
-            if (leader != null && queueSizes != null) {
+
+            Metadata.LeaderAndEpoch leaderAndEpoch = metadata.currentLeader(part);
```
Since the `leaderAndEpoch` variable isn't reassigned in this method, can we grab a reference to the leader `Node` as we had before? That way we don't have quite so many places that we have the more verbose `leaderAndEpoch.leader.isPresent` and `leaderAndEpoch.leader.get()` everywhere.

kirktrue replied:
```suggestion
            Metadata.LeaderAndEpoch leaderAndEpoch = metadata.currentLeader(part);
            Node leader = leaderAndEpoch.orElse(null);
```

Or something like that?

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -842,22 +845,31 @@ private boolean shouldStopDrainBatchesForPartition(ProducerBatch first, TopicPar
             if (isMuted(tp))
                 continue;
-
+            Metadata.LeaderAndEpoch leaderAndEpoch = metadata.currentLeader(new TopicPartition(part.topic(), part.partition()));
```
nit: can we reuse `tp` instead of creating a new `TopicPartition`?

kirktrue reviewed (approved):
I have some minor nits, but I'd be OK as is.

msn-tldr committed:
further address comments

wcarlson5 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -94,9 +100,42 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
+     * It will return true, for a when batch is being retried, it will be retried to a newer leader.
+     */
+
+    boolean hasLeaderChangedForTheOngoingRetry() {
```
love this, thanks!

wcarlson5 reviewed (approved):
Looking good! Thanks for the refactoring bit.

msn-tldr commented:
Looking through CI failures - 
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14384/8/pipeline/10/

For JDK 11 - The 2 flaky runs, and are unrelated to Java clients.
https://ge.apache.org/s/q63xej2ew4w6u

For JDK 17 - Again flaky tests, unrelated to Java client.
https://ge.apache.org/s/dayodofvlspty/tests/overview?outcome=FAILED,FLAKY

ijuma commented:
Would it make sense to backport this to 3.6 as well?

msn-tldr commented:
@ijuma this is merged in 3.6 now, so should be available in 3.6.1.
https://github.com/apache/kafka/pull/14522

ijuma commented:
Thanks!
